File: auth-data-grpc-client/src/main/java/com/cartobucket/auth/data/services/grpc/client/mappers/ScopeMapper.java
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.data.services.grpc.client.mappers;

import com.cartobucket.auth.data.domain.Scope;
import com.cartobucket.auth.data.services.grpc.mappers.MetadataMapper;
import com.cartobucket.auth.rpc.Metadata;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.util.UUID;

public class ScopeMapper {
    public static com.cartobucket.auth.data.domain.Scope toScope(com.cartobucket.auth.data.rpc.Scope scopeResponse) {
        var scope = new com.cartobucket.auth.data.domain.Scope();
        // TODO: This is obviously not right
        scope.setId(!scopeResponse.getId().equals("null") ? UUID.fromString(scopeResponse.getId()) : UUID.randomUUID());
        scope.setName(scopeResponse.getName());
        // TODO: This too is not correct.
        final var authorizationServer = new com.cartobucket.auth.data.domain.AuthorizationServer();
        authorizationServer.setId(!scopeResponse.getAuthorizationServerId().equals("null") ? UUID.fromString(scopeResponse.getAuthorizationServerId()) : UUID.randomUUID());
        scope.setAuthorizationServer(authorizationServer);
        scope.setMetadata(MetadataMapper.from(scopeResponse.getMetadata()));
        scope.setCreatedOn(OffsetDateTime.ofInstant(Instant.ofEpochSecond(scopeResponse.getCreatedOn().getSeconds()), ZoneId.of("UTC")));
        scope.setUpdatedOn(OffsetDateTime.ofInstant(Instant.ofEpochSecond(scopeResponse.getUpdatedOn().getSeconds()), ZoneId.of("UTC")));
        return scope;
    }

    public static com.cartobucket.auth.data.domain.Scope to(String scope) {
        var scopeObj = new com.cartobucket.auth.data.domain.Scope();
        scopeObj.setName(scope);
        return scopeObj;
    }

    public static com.cartobucket.auth.data.rpc.Scope toResponse(Scope scope) {
        // TODO: Fix this too.
        return com.cartobucket.auth.data.rpc.Scope.newBuilder()
                .setId(scope.getId() != null ? scope.getId().toString() : String.valueOf(UUID.randomUUID()))
                .setName(scope.getName())
                .setAuthorizationServerId(String.valueOf(scope.getAuthorizationServer().getId()))
                .setMetadata(scope.getMetadata() != null ? MetadataMapper.to(scope.getMetadata()) : Metadata.newBuilder().build())
                .build();
    }
}

File: web/src/main/kotlin/com/revethq/auth/web/api/routes/mappers/ScopeMapper.kt
/*
 * Copyright 2023 Bryce Groff (Revet)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.revethq.auth.web.api.routes.mappers

import com.revethq.auth.core.domain.AuthorizationServer
import com.revethq.auth.core.api.dto.ScopeRequest
import com.revethq.auth.core.api.dto.ScopeResponse
import com.revethq.auth.core.domain.Scope

object ScopeMapper {

    @JvmStatic
    fun to(scopeRequest: ScopeRequest): Scope {
        val scope = Scope()
        scope.name = scopeRequest.name
        val authorizationServer = AuthorizationServer()
        authorizationServer.id = scopeRequest.authorizationServerId
        scope.authorizationServer = authorizationServer
        scope.metadata = MetadataMapper.from(scopeRequest.metadata)
        return scope
    }

    @JvmStatic
    fun toResponse(scope: Scope): ScopeResponse {
        return ScopeResponse().apply {
            id = scope.id.toString()
            authorizationServerId = scope.authorizationServer?.id
            name = scope.name
            metadata = MetadataMapper.to(scope.metadata)
            createdOn = scope.createdOn
            updatedOn = scope.updatedOn
        }
    }

    @JvmStatic
    fun toSummaryResponse(scope: Scope): ScopeResponse {
        return ScopeResponse().apply {
            id = scope.id.toString()
            name = scope.name
        }
    }
}
